Sources/GRPCCore/GRPCClient.swift

```diff
@@ -0,0 +1,823 @@
+    /// is terminal.
+    case stopped
+  }
+
+  /// Underlying storage for the client.
+  private struct Storage {
+    var interceptors: Interceptors
+    var methodConfigurationOverrides: MethodConfigurationRegistry
+    var state: State
+
+    init() {
+      self.interceptors = Interceptors()
+      self.methodConfigurationOverrides = MethodConfigurationRegistry()
+      self.state = .notStarted
+    }
+  }
+
+  private let storage: LockedValueBox<Storage>
+
+  /// The transport which provides a bidirectional communication channel with the server.
+  private let transport: ClientTransport
+
+  /// Creates a new client with no resources.
+  ///
+  /// You can add resources to the client via ``interceptors-swift.property`` and
+  /// ``methodConfigurationOverrides-swift.property``, and start the client by calling ``run()``.
+  ///
+  /// - Note: Any changes to resources after ``run()`` has been called will be ignored.
+  ///
+  /// - Parameter transport: The ``ClientTransport`` to be used for this ``GRPCClient``.
+  public init(transport: ClientTransport) {
+    self.storage = LockedValueBox(Storage())
+    self.transport = transport
+  }
+
+  /// Start the client.
+  ///
+  /// This is a long-running task that will return once ``close()`` has been called and all in-flight RPCs
+  /// finished executing.
+  ///
+  /// If you need to immediately stop all work, cancel the task executing this method.
+  public func run() async throws {
+    try self.storage.withLockedValue { storage in
+      switch storage.state {
+      case .notStarted:
+        storage.state = .running
+      case .running:
+        throw ClientError(
+          code: .clientIsAlreadyRunning,
+          message: "The client is already running and can only be started once."
+        )
+      case .stopping, .stopped:
+        throw ClientError(
+          code: .clientIsStopped,
+          message: "The client has stopped and can only be started once."
+        )
+      }
+    }
+
+    // When we exit this function we must have stopped.
+    defer {
+      self.storage.withLockedValue { $0.state = .stopped }
+    }
+
+    try await withThrowingTaskGroup(of: Void.self) { group in
+      group.addTask {
+        try await self.transport.connect(lazily: false)
+      }
+      try await group.next()
+    }
+  }
+
+  /// Close the client.
+  ///
+  /// The transport will be closed: this means that it will be given enough time to wait for in-flight RPCs to
+  /// finish executing, but no new RPCs will be accepted.
+  /// You can cancel the task executing ``run()`` if you want to immediately stop all work.
+  public func close() {
+    self.storage.withLockedValue { storage in
+      switch storage.state {
+      case .notStarted:
+        storage.state = .stopped
+      case .running:
+        storage.state = .stopping
+      case .stopping, .stopped:
+        ()
+      }
+    }
+
+    self.transport.close()
+  }
+
+  /// Start a unary RPC.
```

Comment:
This doesn't just start the RPC, it finishes it too. We can just say "executes" though.